### ISSUE TITLE
hide latest tag from image streams

### ISSUE THREAD

**Issue by tmds** [redhat-developer/s2i-dotnetcore#336]
Using `latest` for .NET Core seems an anti-pattern because applications don't work across major versions automatically. Should we hide the `latest` tag from our imagestreams?

@omajid @aslicerh what do you think?

cc @bparees @yselkowitz @gabemontero

**Comment by yselkowitz** [comment]
Incompatibilities are always possible between language runtime versions, yet we give users the risk-documented option of latest anyway.  But depends what you mean by "won't work automatically".  Changes to the image underlying a given imagestream tag trigger a rebuild of the application.  Would some applications built with 3.1 cleanly rebuild with 5.0?  What about from 5.0 to a possible future 5.1?

**Comment by yselkowitz** [comment]
That would be somewhat different from the SCLs where there is a risk of incompatibility but still a good chance that at least some applications will still build and run from one to the next.  Marking latest as hidden here might be appropriate then, but I'll defer to @gabemontero.

**Comment by gabemontero** [comment]
In general I believe I concur with what @tmds is getting at, and is a pattern we see with many of the sample imagestream providers, but I'm going to reiterate what was said with some additional detail to make sure we are on the same page.

1) we need to distinguish between an `ImageStreamTag` of `latest` and an image reference that employs the tag of `latest`.

2) quick reminder: `ImageStreamTag`  in practice either ref another `ImageStreamTag` or an actual image, per the ref type `DockerImage`

3) it is OK to have an `ImageStreamTag` of latest that has an `ImageStreamTag` ref to whatever `ImageStreamTag` has as a `DockerImage` ref to the latest version of your image

4) but any  `ImageStreamTag` with a `DockerImage` ref is better served to have a tag that is not `latest`, but a specific version

the reasons being what @tmds has articulated

**Comment by bparees** [comment]
> it is OK to have an ImageStreamTag of latest that has an ImageStreamTag ref to whatever ImageStreamTag has as a DockerImage ref to the latest version of your image

it's ok, but still sets users up for similar failure modes.  Whether it's a pullspec:latest or imagestream:latest, both have the same implications:  "If you use this, the underlying image you will get may suddenly change in substantially incompatible ways"  (whether because someone pushes a new major version to pullspec:latest, or because someone updates imagestream:latest to ref a new imagestream:majorversion)

We ship "imagestream:latest" tags which ref imagestream:latestMajorVersion (which in turn refs pullspec:latestMajorVersion) for the convenience of someone who just wants to ensure they are on the latest available version right now and aren't concerned with future compatibility.  For dev purposes, it's likely fine.  For production usage, it is not.

**Comment by yselkowitz** [comment]
Like the SCLs, dotnet has different image names for each version, so e.g. ubi8/dotnet-31:latest should NOT be changing in incompatible ways, and only the latest release for each runtime+version pair is supported.

As for dotnet:latest, if you look at it from a developer perspective as stated above, then if they choose :latest, then they're essentially saying they don't mind (iow actually want) to break and update their code e.g. when 5.0 is released, and when they do so, all they have to do is trigger a new build.  Otherwise, they'd have to manually change their Deployment (or DC) to move to the next runtime.

So basically, :latest is for developers, and they're already warned about the risks thereof.  I'd say leave it alone then.

**Comment by tmds** [comment]
> For dev purposes, it's likely fine.

It's less usable than what you might assume. The `dotnet:latest` version will still need to match with the app.
When a developer has an app, he knows the version.
And if he wants to develop something new, the 'latest' version will be decided when the source code is created (for example, because the latest SDK on the developer system is using that version in its template).

There may be use-cases for `dotnet:latest` but they are not common.

**Comment by aslicerh** [comment]
So, maybe an odd question here, but is there really a point to having the 'dotnet' and 'dotnet-runtime' tag lists anymore? Without the ability to specify 'latest', you always end up specifying the specific version and style that you want anyway.

**Comment by yselkowitz** [comment]
There is still a point to having an imagestream without latest.

**Comment by tmds** [comment]
And you can still specify `latest` if you want. It just doesn't show up in the OpenShift UI.
